cpp/src/arrow/util/decimal.cc

```diff
@@ -176,37 +177,46 @@ static constexpr int64_t kPowersOfTen[kInt64DecimalDigits + 1] = {1LL,
                                                                   100000000000000000LL,
                                                                   1000000000000000000LL};
 
-static void StringToInteger(const std::string& str, Decimal128* out) {
+// Iterates over data and for each group of kInt64DecimalDigits multiple out by
+// the appropriate power of 10 necessary to add source parsed as uint64 and
+// then adds the parsed value of source.
+static inline void ShiftAndAdd(const char* data, size_t length, Decimal128* out) {
+  static internal::StringConverter<Int64Type> converter;
```

Comment:
This relies on `StringConverter` being stateless (see other comment).

Reply:
Will measure and continue conversation on the first one.